R/target_encoder.R

```diff
@@ -26,9 +30,111 @@
 #' my_train$y,
 #' cat_columns = c("foo"),
 #' prior = 0.5,
-#' min_samples = 1)
-target_encoder <- function(X_train, X_test, y, cat_columns, prior = 0.5, min_samples = 1) {
+#' objective = "regression")
 
-  out <- list(train_processed, test_processed)
+target_encoder <- function(X_train, X_test = NULL, y, cat_columns, prior = 0.5, objective = "regression") {
 
+  # check input of objective
+  if (!objective %in% c("regression", "binary")) {
+    stop("Objective must be regression or binary.")
+  }
+  # check type of input of prior
+  if (!is.numeric(prior)) {
+    stop("The prior input should be numeric.")
+  }
+  # check type of input X_train
+  if (!is.data.frame(X_train)) {
+    stop("Type of X_train must be pd.Dataframe.")
+  }
+  # check if length y equals to length X_train
+  if (length(y) != nrow(X_train)) {
+    stop("Input y must equal to X_train.")
+  }
+
+  # check if X_train contains cat_columns
+  if (any(cat_columns %in% colnames(X_train)) == FALSE) {
+    stop("Column does not exist in the training set.")
+  }
+
+  #check if target variable is numeric for regression objective
+  if (objective == "regression" & !is_double(y)) {
+      stop("Type of target variable must be numeric.")
+  }
+
+   y_new <- y
+
+  # binary objective
+  if (objective == 'binary') {
+    if (y %>% unique() %>% length() != 2) {
+      stop("The target variable must be binary.")
+    }
+    # encode target variable to 1 and 0
+    if (!is_double(y)){
+      y_new <- case_when(y == unique(y)[1] ~ 0,
+                     y == unique(y)[2] ~ 1)
+    }
+
+  }
+
+  train_processed <- X_train
+
+  global_mean <- mean(y_new, na.rm = TRUE)
+
+  if (is.null(X_test)){
+    for(i in seq_along(cat_columns)){
+      column <- cat_columns[i]
+      # calculate target counts for each category and save to dictionary
+      search_table <- train_processed %>%
+        bind_cols(target = y_new) %>%
+        group_by(!!sym(column)) %>%
+        summarize(the_sum = sum(target, na.rm = TRUE),
+                  the_count = n())
+
+      search_table['encodings'] <- (search_table['the_sum'] + prior * global_mean) / (search_table['the_count'] + prior)
+      search_table = search_table %>% dplyr::select( -c(the_sum,the_count))
+      # # encode categorical columns for training dataset
+      train_processed <- left_join(train_processed, search_table, by = column)
+      train_processed[column] <- train_processed['encodings']
+      train_processed <- train_processed %>% dplyr::select(-encodings)
+      }
+    out <- train_processed
+
+    } else{
```

Comment:
```suggestion
    } else {
```

just for tidyverse standards, that's all.